ideascube/library/tests/test_views.py

```diff
@@ -514,3 +514,88 @@ def test_export_book_specimens(staffapp, monkeypatch):
             '{},,,,,{}'.format(specimen2.item.name, epubname),
 
         ]
+
+
+def test_import_book_specimens(staffapp, tmpdir):
+    book1 = BookFactory(isbn='123456', name='my book title')
+    book2 = BookFactory(isbn='234567', name='ﺎﻠﻨﺒﻳ (ﻚﺗﺎﺑ)')
+
+    zip_path = tmpdir.join('specimens.zip')
+    csv_content = '\n'.join([
+        'item,barcode,serial,comments,location,file',
+        'my book title,1234,,,,',
+        'my book title,,,,,1.epub',  # file is missing from the archive
+        'ﺎﻠﻨﺒﻳ (ﻚﺗﺎﺑ),,,,,2.epub',
+        'ﺎﻠﻨﺒﻳ (ﻚﺗﺎﺑ),1234,,,,',     # this barcode already exists
+        'no such book,2345,,,,',     # no item with this title
+    ])
+
+    with zipfile.ZipFile(str(zip_path), mode='x') as zip:
+        zip.writestr('specimens.csv', csv_content.encode('utf-8'))
+        zip.writestr('2.epub', b'Trust me, this is an epub')
+
+    form = staffapp.get(reverse('library:specimen_import')).forms['import']
+    form['source'] = Upload(
+        zip_path.basename, zip_path.read_binary(), 'application/zip')
+    response = form.submit()
+    assert response.status_int == 302
+    assert response.location.endswith(reverse('library:index'))
+
+    response = response.follow()
+    specimens = BookSpecimen.objects.order_by('id')
+    assert specimens.count() == 2
+
+    specimen = specimens[0]
+    assert specimen.item.pk == book1.pk
+    assert specimen.barcode == '1234'
+    assert specimen.serial is None
+    assert specimen.comments == ''
+    assert specimen.location == ''
+    assert not specimen.file
+
+    specimen = specimens[1]
+    assert specimen.item.pk == book2.pk
+    assert specimen.barcode is None
+    assert specimen.serial is None
+    assert specimen.comments == ''
+    assert specimen.location == ''
+    assert specimen.file.read() == b'Trust me, this is an epub'
+
+    # Errors
+    assert (
+        'Could not import line 2: file 1.epub is missing from the archive'
```

Comment:
I don't see why it wouldn't be.

This isn't even a private key at all, it is just the file name in the tarball.

It just so happens that when **exporting** the file is called after the private key of the exported model instance.

But when **importing**, the new model instance will have its own private key (if successfully imported), which might or might not end up being the same integer as the one used for the file name.

(but if it were a private key, then yes, it would be reliable, since the test gets a clean database with no specimens in it, the first one will therefore be `1`)